mfe/mfe.go

```diff
@@ -0,0 +1,1209 @@
+		return 0, nil, err
+	}
+
+	pairTable, err := pairTable(structure)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	fc := &foldCompound{
+		length:              lenSequence,
+		energyParams:        scaleEnergyParams(temperature),
+		sequence:            sequence,
+		encodedSequence:     encodeSequence(sequence),
+		pairTable:           pairTable,
+		basePairEncodedType: basePairEncodedTypeMap(),
+	}
+
+	energyInt, energyContributions := evaluateFoldCompound(fc)
+	energy := float64(energyInt) / 100.0
+
+	return energy, energyContributions, nil
+}
+
+func ensureValidRNA(sequence string) error {
+	rnaRegex, _ := regexp.Compile("^[ACGU]+")
+	rnaIdxs := rnaRegex.FindStringIndex(sequence)
+
+	if rnaIdxs[0] != 0 || rnaIdxs[1] != len(sequence) {
+		return fmt.Errorf("found invalid characters in RNA sequence. Only A, C, G, and U allowed")
+	}
+	return nil
+}
+
+func ensureValidStructure(structure string) error {
+	dotBracketStructureRegex, _ := regexp.Compile("^[().]+")
+	structureIdxs := dotBracketStructureRegex.FindStringIndex(structure)
+
+	if structureIdxs[0] != 0 || structureIdxs[1] != len(structure) {
+		return fmt.Errorf("found invalid characters in structure. Only dot-bracket notation allowed")
+	}
+	return nil
+}
+
+// Encodes a sequence into its numerical representaiton
+func encodeSequence(sequence string) []int {
+	// Make a nucleotide byte -> int map
+	nucleotideRuneMap := map[byte]int{
+		'A': 1,
+		'C': 2,
+		'G': 3,
+		'U': 4,
+	}
+
+	lenSequence := len(sequence)
+	encodedSequence := make([]int, lenSequence)
+
+	// encode the sequence based on nucleotideRuneMap
+	for i := 0; i < lenSequence; i++ {
+		encodedSequence[i] = nucleotideRuneMap[sequence[i]]
+	}
+
+	return encodedSequence
+}
+
+/**
+* Returns a map that encodes a base pair to its numerical representation.
+* The map is:
+*    _  A  C  G  U
+*	_ {0, 0, 0, 0, 0}
+*	A {0, 0, 0, 0, 5}
+*	C {0, 0, 0, 1, 0}
+*	G {0, 0, 2, 0, 3}
+*	U {0, 6, 0, 4, 0}
+* For example, map['A']['U'] is 5.
+* The encoded numerical representation of a base pair carries no meaning in
+* itself except for where to find the relevent energy contributions of the base
+* pair in the matrices of the energy paramaters (found in `energy_params.go`).
+* Thus, any change to this map must be reflected in the energy
+* parameters matrices in `energy_params.go`, and vice versa (see
+* `energy_params.md` for more information.)
+ */
+func basePairEncodedTypeMap() map[byte]map[byte]int {
+	nucelotideAEncodedTypeMap := map[byte]int{'U': 5}
+	nucelotideCEncodedTypeMap := map[byte]int{'G': 1}
+	nucelotideGEncodedTypeMap := map[byte]int{'C': 2, 'U': 3}
+	nucelotideUEncodedTypeMap := map[byte]int{'A': 6, 'G': 4}
+
+	basePairMap := map[byte]map[byte]int{
+		'A': nucelotideAEncodedTypeMap,
+		'C': nucelotideCEncodedTypeMap,
+		'G': nucelotideGEncodedTypeMap,
+		'U': nucelotideUEncodedTypeMap,
+	}
+
+	return basePairMap
+}
+
+// Contains all the energy parameters needed for the free energy calculations
+type energyParams struct {
```

Comment:
parameterize as many magic numbers as possible in here.

Reply:
What do you mean by parameterize?

Reply:
Give them variable assignments. numbers should be assigned to variables.